KAFKA-8058: Fix ConnectClusterStateImpl.connectors() method

C0urante opened the pull request:
This makes the `ConnectClusterStateImpl.connectors()` method synchronous, whereas before it was implicitly asynchronous with no way to tell whether it had completed or not.

More detail can be found in the [Jira](https://issues.apache.org/jira/browse/KAFKA-8058).

~Tested manually. Seems like a light enough change that even unit tests would be overkill, but if reviewers feel differently tests can be added.~
A unit test has also been added for the `ConnectClusterStateImpl` class.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

C0urante committed:
KAFKA-8058: Fix ConnectClusterStateImpl.connectors() method

C0urante commented:
@mageshn @wicknicks @rayokota could you guys take a look when you have a chance? 😃

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java:
```diff
@@ -42,13 +45,22 @@ public ConnectClusterStateImpl(HerderProvider herderProvider) {
+        try {
+            // Should take no more than 10 seconds to get a list of connectors. Any longer than this
+            // and it's very likely something is wrong.
+            connectorsLatch.await(10, TimeUnit.SECONDS);
```
nit: should we log an error if this method returns false?

C0urante replied:
Hmm... good point. I think it may make more sense raise a `ConnectException` in order to avoid incorrect/unexpected behavior from the code that's invoking this method. WDYT? Would be interested in @mageshn's thoughts as well

rayokota replied:
Yes, it would probably be good to throw an exception here if the method returns false.

wicknicks reviewed (commented):
good catch, @C0urante. one comment.

mageshn reviewed (approved):
Nice catch @C0urante. LGTM.

C0urante committed:
KAFKA-8058: Change ConnectorClusterStateImpl.connectors() to throw exception on timeout

rayokota reviewed (approved):
LGTM

wicknicks reviewed (approved):
Thanks, @C0urante. LGTM.

C0urante commented:
@rhauch @ewencp Hi guys! Got a minor bug fix for you; been reviewed and approved by three contributors and seems ready to be looked at by a committer or two. Would appreciate your review when you get the chance.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java:
```diff
@@ -42,13 +45,25 @@ public ConnectClusterStateImpl(HerderProvider herderProvider) {
             @Override
             public void onCompletion(Throwable error, Collection<String> result) {
                 connectors.addAll(result);
+                connectorsLatch.countDown();
```
@C0urante would using a `FutureCallback` be more appropriate here? 
Also do you happen to know what happens if this request is not forwarded to the leader of the group? How much do we care here (not easy to say because this class is missing javadoc)?

You might want to check `completeOrForwardRequest` to see what we do on another call to `herder.connectors` method elsewhere.

C0urante replied:
@kkonstantine thanks for pointing out the `FutureCallback` class, that's much better than the `CountDownLatch`-based approach.

The `ConnectClusterStateImpl` class doesn't have javadocs, but the `ConnectClusterState` interface does. There's no mention of expected behavior if the `ConnectClusterState.connectors()` method is invoked on or off the master node, but FWICT both the standalone and distributed herders rely on their local `ClusterConfigState` snapshots to provide the list of connectors for their own `connectors()` methods, so it doesn't appear forwarding to the leader will be necessary. Let me know if I'm missing anything here; I've tried to go over the code base as best as I can but it's always good to hear from someone more familiar with it.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java:
```diff
@@ -42,13 +45,25 @@ public ConnectClusterStateImpl(HerderProvider herderProvider) {
+        try {
+            // Should take no more than 10 seconds to get a list of connectors. Any longer than this
+            // and it's very likely something is wrong.
+            boolean success = connectorsLatch.await(10, TimeUnit.SECONDS);
```
Have you compared this value with `ConnectorsResource.REQUEST_TIMEOUT_MS`? I know the other one is taking into account rebalances. But how much do we care here? The diff between 90sec and 10sec seems significant.

C0urante replied:
It looks like the `DistributedHerderRequest` that gets created to fulfill a call to `DistributedHerder.connectors(...)` should complete quickly and won't be blocked even during rebalance. I'm unclear if other requests that are queued up before it could cause a call to `DistributedHerder.connectors(...)` to still take longer than ten seconds but I'm leaning toward no...

kkonstantine replied:
Timeout to read the log and get a snapshot is 3sec. I agree. I think we are good.

kkonstantine reviewed (commented):
Nice find @C0urante !
I think using `FutureCallback` already allows you to do what you implement with the countdown latch. 
Also I have a question on whether we need to forward this request to the leader as we do elsewhere.

kkonstantine commented:
Forgot to mention that a unit test would be nice (for the whole class even?), since this seems it never worked because it returned immediately.

C0urante committed:
KAFKA-8058: Address review comments, add unit test for ConnectClusterStateImpl.connectors()

C0urante commented:
@kkonstantine I think I've addressed your two inline comments, and I've added a unit test for the `ConnectClusterStateImpl.connectors()` method. The other method (`connectorHealth(...)`) is synchronous and doesn't really look like it warrants an additional test.

As far as the ten-second timeout goes, I'm open to suggestions but it doesn't appear that rebalances or forwarding to master should be a factor.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java:
```diff
@@ -0,0 +1,120 @@
+  
+    protected static final long CONNECTORS_TIMEOUT_MS = 2_000;
+    
+    protected Herder mockHerder;
```
nit: my impression is that we don't name the mocks as `mockSomething`. We just declare them/initialize them. Also, I think the annotation style is preferable with easymock/powermock, but I don't feel strong about it.

C0urante committed:
KAFKA-8058: Simplify unit tests for ConnectClusterStateImpl.connectors()

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java:
```diff
@@ -0,0 +1,85 @@
+    }
+    
+    @Test
+    public void connectors() {
```
Thanks for simplifying the tests @C0urante! Definitely on the right track. 
As you can see in https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java 
you can use the already existing `TestFuture` and be a bit more specific on what to capture and what you expect. 

For example this could be written as: 
```java
    @Test
    public void connectors() {
        Capture<Callback<Collection<String>>> callback = EasyMock.newCapture();
        mockHerder.connectors(EasyMock.capture(callback));
        EasyMock.expectLastCall().andAnswer(new IAnswer<Future<Collection<String>>>() {
            @Override
            public Future<Collection<String>> answer() throws Throwable {
                TestFuture<Collection<String>> future = new TestFuture<>();
                callback.getValue().onCompletion(null, expectedConnectors);
                future.resolveOnGet(expectedConnectors);
                return future;
            }
        });
        EasyMock.replay(mockHerder);
        assertEquals(expectedConnectors, connectClusterState.connectors());
    }
```
this way you also manage to avoid unchecked casts.

C0urante replied:
@kkonstantine definitely prefer the `Capture`-based approach that you've outlined, but I don't think the `TestFuture`/`IAnswer` combination in your example is applicable since the future/callback we're dealing with is being passed to `Herder.connectors(...)`, instead of being returned by it.

kkonstantine replied:
That's correct. It's called in the method you test. I was carried away in the effort to remove the unchecked casts.

kkonstantine reviewed (commented):
@C0urante this looks even better now IMO and on par with the rest of the code. 
One nit and one more suggested improvement on the tests. Thanks for including those btw. Definitely on the right track.

C0urante committed:
KAFKA-8058: Address review comments on ConnectClusterStateImplTest

C0urante commented:
@kkonstantine thanks for the review! I've taken your comments into account and pushed the next iteration, which consists of changes to how the `ConnectClusterStateImpl` is tested. Hope this looks good and addresses your review remarks, but will be happy to keep iterating if not or if there's anything else that needs work.

kkonstantine reviewed (approved):
LGTM!

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java:
```diff
@@ -17,22 +17,27 @@
 
 public class ConnectClusterStateImpl implements ConnectClusterState {
+    
+    private static final long CONNECTORS_TIMEOUT_MS = 10_000;
```
Nit: What about `TimeUnit.SECONDS.toMillis(10)`?

rhauch reviewed (commented):
Nice find, @C0urante! One really minor suggestion for slight improvement in readability, but not a bit deal.

This has never worked, right? This PR should be relatively easy to backport.

C0urante committed:
KAFKA-8058: Address review comments

C0urante commented:
@rhauch yep, this bug has been present since REST extensions were added to Connect. I've added your suggested change to the PR, let me know if there are any other alterations you'd like.

rhauch reviewed (approved):
Thanks, @C0urante! LGTM.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java:
```diff
@@ -0,0 +1,84 @@
+        assertEquals(expectedConnectors, connectClusterState.connectors());
+    }
+
+    @Test(expected = ConnectException.class)
```
We're moving away from using this since we can use the new `assertThrows` that junit 4.13 supports, which is better since it isolates to the specific call that should fail and allows you to capture and make further assertions. At least for any new tests, we should prefer `assertThrows`, though that does affect ability to backport (since https://github.com/apache/kafka/pull/6248 was only merged to `master`, I think after `2.2` was cut).

ewencp reviewed (commented):
Mostly looks good, one comment. This is the sort of thing that should come with list of backports -- the bug appears to already be tagged with at least some info about previous versions that are also affected so we probably don't want to close without backporting.

Regarding timeout, it should really be as conservative as possible without being longer than the minimum of (rebalance timeout, default REST API request timeout).

C0urante commented:
@ewencp can you shed some light on why we'd want to take rebalances into account for the timeout? I could be wrong, but I don't think a rebalance would affect the latency of a call to `Herder.connectors(...)`.

C0urante committed:
KAFKA-8058: Use assertThrows(...) instead of expected field in @Test annotation

ewencp commented:
Calling `connectors()` for `DistributedHerder` adds a request to the request queue, which are processed serially in the thread that manages cluster membership so they get a consistent view of the state of the Herder. Pretty much all the requests follow that pattern, though it is true some could potentially be optimized to not require it if some other locking were put in place for certain parts of the state.

The thread that processes the requests is also the one managing the `member` and calling `poll()` on it, which can block on ensuring membership in the group, which means request processing latency can be affected by rebalances.

C0urante committed:
KAFKA-8058: Make timeout for ConnectClusterStateImpl.connetors() more conservative

C0urante commented:
@ewencp thanks for the clarification; I did not realize that membership and request handling were taken care of with the same thread. I've taken your suggestion literally and just made the timeout the minimum of the timeout for the `/connectors` resource and the rebalance timeout (if running in distributed mode). LMK how this looks!

C0urante commented:
BTW, I think we want to backport this to 2.0, which according to [the KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-285%3A+Connect+Rest+Extension+Plugin) is when Connect REST extensions were introduced and AFAICT the code for the `ConnectClusterStateImpl` class hasn't been modified since.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java:
```diff
@@ -308,10 +309,16 @@ void registerRestExtensions(HerderProvider provider, Plugins plugins, ResourceCo
             config, ConnectRestExtension.class);
 
+        long herderRequestTimeoutMs = ConnectorsResource.REQUEST_TIMEOUT_MS;
+        if (config instanceof DistributedConfig) {
```
nit: i don't love `instanceof` solutions. might be better to generalize thee rebalance timeout so support a small/zero value for `StandaloneConfig`.

I wouldn't block merging on this, but this is an anti-pattern, so if there is a cleaner solution, it might be worth exploring.

rhauch replied:
@C0urante, what do you think about doing something like the following instead of the `instanceof` block?

```
        Integer rebalanceTimeoutMs = config.getInt(DistributedConfig.REBALANCE_TIMEOUT_MS_CONFIG);
        if (rebalanceTimeoutMs != null) {
            herderRequestTimeoutMs = Math.min(herderRequestTimeoutMs, rebalanceTimeoutMs.longValue());
        }
```
This still uses the `DistributedConfig.REBALANCE_TIMEOUT_MS_CONFIG`, but otherwise is simply checking for the existence of that property to determine if `herderRequestTimeoutMs` should be set to a new value.

C0urante replied:
@rhauch sgtm! I've added your code snippet verbatim to the PR

ewencp reviewed (approved):
one comment that we could follow up on, but i'm fine merging as is

C0urante committed:
Address review suggestions
